alarm.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH alarm 1 "August 12, 2001"
.SH NAME
alarm \- run a program with a timeout
.SH SYNOPSIS
.B alarm
.BI [ options ]
.RB [ -- ]
.I timeout
.IR command ...
.SH DESCRIPTION
.B alarm
runs a program as a subprocess, and sends it a signal when a timeout
expires (if the program does not exit first).
.PP
The \fItimeout\fR argument may be just a number of seconds, or it may
consist of any or all of seconds, minutes, hours and days.
.PP
In the latter form, the number of seconds should be suffixed by "s",
the number of minutes by "m", the numbers of hours by "h" and the
number of days by "d".  The different components may appear in any
order, but if the same size units are used more than once, the meaning
is not defined.
.PP
Letter case is not significant.  No whitespace is allowed in the
time specification.  All numbers are in decimal.
.PP
The command is passed directly to \fBexecvp\fR(3), rather than
re-invoking the shell.  If you want to execute an arbitrary shell
command it is up to you to invoke the shell properly.
.SH OPTIONS
.TP
\fB-s\fR \fIsignal\fR, \fB--signal\fR \fIsignal\fR
Sets the signal to send to the subprocess when the timeout expires.
You may specify a signal number (e.g. \fB-s3\fR) or a name with or
without the initial "SIG" (e.g. \fB-sSIGINT\fR or \fB-sINT\fR).
Letter case is not significant.
.IP
The default is SIGTERM.
.TP
\fB-v\fR, \fB--verbose\fR
Verbose mode.  If this option is used then \fBalarm\fR will display a
message when it times out the child process, and report the exit
status of the child process (unless it is 0).  If it used twice then
it will report the exit status even when it is 0.
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH "EXIT STATUS"
If the child process exits normally, its status is propagated back to
the caller.  If it terminates due to a signal (whether sent by
\fBalarm\fR or not) then \fBalarm\fR will exit with a status equal to
the signal number plus 128.
.PP
If an error occurs inside \fBalarm\fR, it will exit with status 1.
.SH OUTPUT
\fBalarm\fR reports various conditions, not all of them fatal errors,
to stderr.  It only writes to stdout if the \fB--help\fR or
\fB--version\fR options are used.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>.


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



alarm.c
/*

   alarm - run a command with a timeout

   Copyright (C) 2001 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>
#include <getopt.h>
#include <unistd.h>
#include <stdio.h>
#include <limits.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <stdlib.h>
#include <sys/wait.h>

#include "utils.h"

static int verbose;			/* verbose mode */

/* Option flags and variables */

static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { "signal", required_argument, 0, 's' },
  { "verbose", no_argument, 0, 'v' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fprintf(fp, "Usage:\n"
	     "\n"
	     "  alarm [options] -- timeout command ...\n"
	     "\n"
	     "Options:\n"
	     "\n"
	     "  -s SIGNAL, --signal SIGNAL       Signal to send program on expiry\n"
	     "  -v, --verbose                    Verbose mode\n"
	     "\n") < 0)
    fatale("output error");
  exit(exit_status);
}

/* return a+p*q - supposed to do range checking too, but we don't
 * bother yet! */

static int mla(int a, int p, int q) {
  return a + p * q;
}

/* parse a time specification and return its meaning in seconds */

static int parse_timeout(const char *s) {
  const char *str = s;
  int n = 0;
  int c;
  int t = 0;
  int l = 0;

  while((c = (unsigned char)*s++)) {
    int scale;

    if(isdigit(c)) {
      t = 10 * t + c - '0';
      ++l;
      continue;
    }
    switch(c) {
    case 's':
    case 'S':
      scale = 1;
      break;
    case 'm':
    case 'M':
      scale = 60;
      break;
    case 'h':
    case 'H':
      scale = 3600;
      break;
    case 'd':
    case 'D':
      scale = 86400;
      break;
    default:
      fatal("unknown time unit '%c' in '%s'", c, str);
    }
    if(!l)
      fatal("invalid time specification '%s'", str);
    n = mla(n, t, scale);
    l = 0;
    t = 0;
  }
  n = mla(n, t, 1);
  return n;
}

/* convert a signal name into a number */

static int parse_signal(const char *s) {
  const char *str = s;
  int n;

  if(strspn(s, "0123456789") == strlen(s))
    return atoi(s);
  /* the initial SIG is optional */
  if(!strncasecmp(s, "SIG", 3))
    s += 3;
  if((n = lookupi(signallookup, s)) >= 0)
    return n;
  fatal("unrecognized signal '%s'", str);
}

static pid_t pid;
static int child_killer_signal = SIGTERM;

static void alarm_handler(int __attribute__((unused)) sig) {
  /* timed out */
  if(verbose)
    error("timing out child process");
  if(kill(-pid, child_killer_signal) < 0)
    fatale("error sending signal %d to process group %lu",
	   child_killer_signal, (unsigned long)pid);
  /* let child_handler tidy up the child */
}

static void child_handler(int __attribute__((unused)) sig) {
  /* child terminated */
  int w;
  pid_t r;

  r = waitpid_e(pid, &w, WNOHANG);
  if(r == pid) {
    if(WIFEXITED(w)) {
      if((w == 0 && verbose > 1)
	 || (w != 0 && verbose > 0))
	fprintf(stderr, "%s: child %s\n", program_name, wstat(w));
      exit(w);
    } else {
      if(verbose)
	fprintf(stderr, "%s: child %s\n", program_name, wstat(w));
      exit(WTERMSIG(w) + 128);
    }
  }
}

int main(int argc, char **argv) {
  int n;
  int timeout;

  /* shut off any existing alarm, it will merely confuse matters */
  alarm(0);
  setprogname(argv[0]);

  while((n = getopt_long(argc, argv,
			 "hVs:v",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("alarm %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);
      return 0;

    case 's':
      child_killer_signal = parse_signal(optarg);
      break;

    case 'v':
      ++verbose;
      break;
      
    default:
      usage(stderr, 1);
    }

  }

  /* get the timeout */
  if(optind >= argc)
    fatal("no timeout speciifed");
  timeout = parse_timeout(argv[optind++]);

  if(optind >= argc)
    fatal("no command specified");

  /* set up signals */
  handle_signal(SIGALRM, alarm_handler);
  handle_signal(SIGCHLD, child_handler);

  /* actually install the handlers.  This needs to happen before the
   * fork as otherwise the signal mask may let an unhandled SIGCHLD
   * through after the fork but before we install a handler.  We'll
   * call restore_signals() inside the fork to put things back
   * right. */
  init_signals();
  
  /* create the child process */
  if(!(pid = fork_e())) {
    exiter = _exit;
    restore_signals();
    /* set our process group */
    if(setpgid(0, 0) < 0)
      fatale("error calling setpgrp");
    /* execute the command */
    execvp(argv[optind], argv + optind);
    fatale("error executing %s", argv[optind]);
  }

  /* Set the child's process group.  We do this in both the child so
   * that if the child forks again after the exec but before the
   * parent gets to run again the process group is right, and in the
   * parent so that if the parent gets to run first and sends a signal
   * to the child, the process group is already set. */
  if(setpgid(pid, pid) < 0)
    fatale("error calling setpgid");

  /* set the alarm */
  alarm(timeout);

  /* wait for signals to come in */
  signal_loop();
 
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
